DigitalLearningSolutions.Web/Controllers/NotificationPreferencesController.cs

```diff
@@ -46,30 +46,48 @@ public IActionResult Index()
         [Authorize]
         [HttpGet]
         [Route("/NotificationPreferences/Edit/{userType}")]
-        public IActionResult UpdateNotificationPreferences(string userType)
+        public IActionResult UpdateNotificationPreferences(UserType? userType)
         {
-            var userId = ((UserType)userType).Equals(UserType.AdminUser)
-                ? User.GetAdminId()
-                : User.GetCandidateIdKnownNotNull();
-            var notifications = notificationPreferencesService.GetNotificationPreferencesForUser(userType, userId);
+            var userId = GetUserId(userType);
+            if (userId == null)
+            {
+                return NotFound();
+            }
 
-            var model = new UpdateNotificationPreferencesViewModel(notifications, userType);
+            var notifications = notificationPreferencesService.GetNotificationPreferencesForUser(userType!, userId);
+
+            var model = new UpdateNotificationPreferencesViewModel(notifications, userType!);
 
             return View(model);
         }
 
         [Authorize]
         [HttpPost]
         [Route("/NotificationPreferences/Edit/{userType}")]
-        public IActionResult SaveNotificationPreferences(string userType, IEnumerable<int> notificationIds)
+        public IActionResult SaveNotificationPreferences(UserType? userType, IEnumerable<int> notificationIds)
         {
-            var userId = ((UserType)userType).Equals(UserType.AdminUser)
-                ? User.GetAdminId()
-                : User.GetCandidateIdKnownNotNull();
+            var userId = GetUserId(userType);
+            if (userId == null)
+            {
+                return NotFound();
+            }
 
-            notificationPreferencesService.SetNotificationPreferencesForUser(userType, userId, notificationIds);
+            notificationPreferencesService.SetNotificationPreferencesForUser(userType!, userId, notificationIds);

```

Comment:
Ah, yeah, so now that you've extracted `GetUserId()` this `!` scares me a bit, because it's relying on the implementation of a method you're calling - someone might come along and change the implementation of `GetUserId()` later and let a null `userType` through.
I think extracting `GetUserId()` is a good shout. And I think you could get around this null check by modifying `GetUserId()` so that it returns a `UserRef?` instead, which should be null if the user doesn't have an account for the `userType` provided. Then you can reject with `NotFound` if the `userRef` is null, and use `userRef.UserType` here, which the compiler will know is non-null.
I think that's worth changing, but I'll pass you onto @stellake - she might have thoughts on this too.